backend/src/main/kotlin/com/ci5644/trade/services/auth/AuthorizationService.kt

```diff
@@ -0,0 +1,103 @@
+import org.springframework.stereotype.Service
+import javax.naming.AuthenticationException
+import org.springframework.security.crypto.password.PasswordEncoder
+
+/**
+ * Service class for handling authorization-related operations.
+ */
+@Service
+class AuthorizationService {
+
+    @Autowired
+    private lateinit var authenticationManager: AuthenticationManager
+
+    @Autowired
+    private lateinit var userRepository: UserRepository
+
+    @Autowired
+    private lateinit var encoder: PasswordEncoder
+
+    /**
+     * Retrieves user entity by username.
+     * @param username String - The username of the user.
+     * @return UserEntity - The user entity.
+     * @throws UsernameNotFoundException if the user with the specified username does not exist.
+     */
+    @Throws(UsernameNotFoundException::class)
+    fun retrieveUser(username: String): UserEntity {
+        if (!userRepository.existsByUsername(username)) {
+            throw UsernameNotFoundException("Username not found: ${username}")
+        }
+        return userRepository.findByUsername(username)
+    }
+
+    /**
+     * Registers a new user.
+     * @param reg RegisterDTO - The registration details of the new user.
+     * @return UserEntity - The newly registered user entity.
+     * @throws UsernameTakenException if the specified username is already taken.
+     */
+    @Throws(UsernameTakenException::class, IllegalArgumentException::class)
+    fun registerUser(reg: RegisterDTO): UserEntity {
+        if (userRepository.existsByUsername(reg.username.trim())) {
+            throw UsernameTakenException()
+        }
+        if (reg.username.length < 5) {
+            throw IllegalArgumentException("Username must be longer or equal than 5 characters")
+        }
+        if (reg.password.length < 8) {
+            throw IllegalArgumentException("Password must be longer or equal than 8 characters")
+        }
+
+        val newUser = UserEntity(
+            username=reg.username.trim(),
+            password=encoder.encode(reg.password),
+            name=reg.name,
+            email=reg.email,
+            gender=reg.gender
+        )
+
+        return userRepository.save(newUser)
+    }
+
+    /**
+     * Logs in a user.
+     * @param username String - The username of the user.
+     * @param password String - The password of the user.
+     * @return HttpCookie - The JWT authentication cookie.
+     * @throws UsernameNotFoundException if authentication fails.
+     */
+    @Throws(UsernameNotFoundException::class, AuthenticationException::class)
+    fun loginUser(username: String, password: String): HttpCookie {
+        try {
+            val appUser = retrieveUser(username.trim())
+            val auth = authenticationManager.authenticate(
+                UsernamePasswordAuthenticationToken(appUser.username, password)
+            )
+            SecurityContextHolder.getContext().authentication = auth
```

Comment:
Extraact this to an util method with  and withtout the lines below. I guarantee you they will be needed in other services